web/src/components/welcome-modal/welcome-modal.tsx

```diff
@@ -30,21 +33,44 @@ export default ({ challengeToken, teamToken, ...props }: WelcomeModalProps) => {
 
   useEffect(() => trackChallenge('modal-welcome'), []);
 
-  const parseEnrollment = (queryString: string): Enrollment => {
-    const regex = new RegExp(/([a-z]+)=([a-z]+)/, 'gm');
+  const parseEnrollment = (
+    queryString: string,
+    referer?: string
+  ): Enrollment => {
+    const regex = new RegExp(/([\w\-]+)=([\w\-]+)/, 'g');
     const queries = {} as { [key: string]: string };
     let pair: Array<string> = [];
 
     while ((pair = regex.exec(queryString)) !== null) {
       queries[pair[1]] = pair[2];
     }
 
-    return {
-      challenge: queries.challenge,
-      team: queries.team,
-      referer: queries.referer,
-      invite: queries.invite,
-    } as Enrollment;
+    queries.referer = referer;
+
+    if (
+      challengeTokens.includes(queries.challenge as ChallengeToken) &&
+      challengeTeamTokens.includes(queries.team as ChallengeTeamToken)
+    ) {
+      return pick(queries, 'challenge', 'team', 'invite') as Enrollment;
+    } else return null;
+  };
+
+  const redirectEnrollment = (enrollmentDetails: string, referrer?: string) => {
+    const referrerString = referrer ? `&referer=${referrer}` : '';
+
+    if (enrollmentDetails) {
+      if (account) {
+        const enrollObject = parseEnrollment(enrollmentDetails, referrer);
+
+        saveAccount({ enrollment: enrollObject }).then(() => {
+          window.location.href = `/dashboard/challenge?challenge=${enrollObject.challenge}&achievement=1${referrerString}`;
+        });
+      } else {
+        window.location.href = `/login${enrollmentDetails}${referrerString}`;
```

Comment:
I tried a bunch of different ways to get these two working with `<Redirect />` and while I could get logged-in users to redirect, I could not get the logged out version to send people to `/login` for the life of me. I'm fine with this the way it is, but if you know how to make `<Redirect />` behave I'm all ears!!

Reply:
Oops, I should have been more specific! I don't think `/login` is handled by React Router, so it needs a redirect as above. But logged-in users should be able to use `<Redirect />` (as you noticed), since dashboard routes are handled on the front-end. I think we should use `<Redirect />` wherever we can.

---

We _could_ handle external routes within our router by following a [pattern like this](https://jessewaites.com/Redirecting-To-An-External-Link-In-A-React-Application), but it's a matter of taste at that point.

Reply:
Very relieved that I wasn't just missing something totally glaring about how `<Redirect />` works 😅